README
NAME
    Mojo::TFTPd - Trivial File Transfer Protocol daemon

VERSION
    0.03

SYNOPSIS
        use Mojo::TFTPd;
        my $tftpd = Mojo::TFTPd->new;

        $tftpd->on(error => sub {
            warn "TFTPd: $_[1]\n";
        });

        $tftpd->on(rrq => sub {
            my($tftpd, $c) = @_;
            open my $FH, '<', $c->file;
            $c->filehandle($FH);
            $c->filesize(-s $c->file);
        });

        $tftpd->on(wrq => sub {
            my($tftpd, $c) = @_;
            open my $FH, '>', '/dev/null';
            $c->filehandle($FH);
        });

        $tftpd->on(finish => sub {
            my($tftpd, $c, $error) = @_;
            warn "Connection: $error\n" if $error;
        });

        $tftpd->start;
        $tftpd->ioloop->start unless $tftpd->ioloop->is_running;

DESCRIPTION
    This module implements a server for the Trivial File Transfer Protocol
    <http://en.wikipedia.org/wiki/Trivial_File_Transfer_Protocol>.

    From Wikipedia:

        Trivial File Transfer Protocol (TFTP) is a file transfer protocol notable
        for its simplicity. It is generally used for automated transfer of
        configuration or boot files between machines in a local environment.

    The connection ($c) which is referred to in this document is an instance
    of Mojo::TFTPd::Connection.

EVENTS
  error
        $self->on(error => sub {
            my($self, $str) = @_;
        });

    This event is emitted when something goes wrong: Fail to "listen" to
    socket, read from socket or other internal errors.

  finish
        $self->on(finish => sub {
            my($self, $c, $error) = @_;
        });

    This event is emitted when the client finish, either successfully or due
    to an error. $error will be an empty string on success.

  rrq
        $self->on(rrq => sub {
            my($self, $c) = @_;
        });

    This event is emitted when a new read request arrives from a client. The
    callback should set "filehandle" in Mojo::TFTPd::Connection or the
    connection will be dropped.

  wrq
        $self->on(wrq => sub {
            my($self, $c) = @_;
        });

    This event is emitted when a new write request arrives from a client.
    The callback should set "filehandle" in Mojo::TFTPd::Connection or the
    connection will be dropped.

ATTRIBUTES
  ioloop
    Holds an instance of Mojo::IOLoop.

  listen
        $str = $self->server;
        $self->server("127.0.0.1:69");
        $self->server("tftp://*:69"); # any interface

    The bind address for this server.

  max_connections
    How many concurrent connections this server can handle. Default to 1000.

  retries
    How many times the server should try to send ACK or DATA to the client
    before dropping the connection.

  inactive_timeout
    How long a connection can stay idle before being dropped.

METHODS
  start
    Starts listening to the address and port set in "Listen". The "error"
    event will be emitted if the server fail to start.

AUTHOR
    Svetoslav Naydenov

    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution Mojo-TFTPd

1.00 Not Released
 - Updated authors and copyright

0.05 2023-03-16T13:22:04
 - Fix sending "\0" at end of OACK
 - Fix cpanfile
 - Update basic repo files

0.04 2014-11-08T05:21:26Z
 - Retransmit last packet on timeout
 - Rollover support
 - Custom Connection class
 - win32 compatibility
 - Support for Mojo::Asset for filehandle

0.03 2014-08-13T16:19:51Z
 - Add RFC 2347 TFTP Option Extension
 - Add RFC 2348 TFTP Blocksize Option
 - Add RFC 2349 TFTP Timeout Interval and Transfer Size Options

0.0203 2013-09-06T23:42:35Z
 - Fix Makefile.PL

0.0202 2013-09-06T10:54:44Z
 - Fix t/basic.t on SunOS/Solaris

0.0201 2013-08-26T23:19:36Z
 - Skip test if on windows
   http://www.cpantesters.org/cpan/report/0714a910-6c03-1014-b581-fa2409253afe
   "Your vendor has not defined Socket macro MSG_DONTWAIT..."

0.02 2013-08-13T18:16:58Z
 - Fixed Mojo::URL regression
 - Fixed 508 byte block RFC-violation Author: https://github.com/hemmop
   https://github.com/jhthorsen/mojo-tftpd/pull/1

0.01 2012-11-11T02:40:10Z
 - Started project


Makefile.PL
use 5.016;
use strict;
use warnings;
use utf8;
use ExtUtils::MakeMaker;

my $GITHUB_URL = 'https://github.com/jhthorsen/mojo-tftpd';

my %WriteMakefileArgs = (
  AUTHOR        => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE       => 'artistic_2',
  NAME          => 'Mojo::TFTPd',
  ABSTRACT_FROM => 'lib/Mojo/TFTPd.pm',
  VERSION_FROM  => 'lib/Mojo/TFTPd.pm',
  EXE_FILES     => [],
  TEST_REQUIRES => {'Test::More'  => '1.30'},
  PREREQ_PM     => {'Mojolicious' => '9.00'},
  META_MERGE    => {
    'dynamic_config' => 0,
    'meta-spec'      => {version   => 2},
    'no_index'       => {directory => [qw(examples t)]},
    'prereqs'        => {runtime   => {requires => {perl => '5.016'}}},
    'resources'      => {
      bugtracker => {web => "$GITHUB_URL/issues"},
      homepage   => $GITHUB_URL,
      license    => ['http://www.opensource.org/licenses/artistic-license-2.0'],
      repository => {type => 'git', url => "$GITHUB_URL.git", web => $GITHUB_URL},
      x_IRC => {url => 'irc://irc.libera.chat/#convos', web => 'https://web.libera.chat/#convos'},
    },
    'x_contributors' => [
      'Per Carlson <pelle@hemmop.com>',
      'Philippe Bruhat (BooK) <book@cpan.org>',
      'Svetoslav Naydenov <harryl@cpan.org>',
    ],
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


example/tftpd.pl
#!/usr/bin/env perl
use Mojo::Base -strict;
use Mojo::TFTPd;

# MOJO_TFTPD_DEBUG=1 perl -Ilib example/tftpd.pl
my $tftpd = Mojo::TFTPd->new(listen => 'tftp://*:12345');

$tftpd->on(
  error => sub {
    warn "Mojo::TFTPd: ", $_[1], "\n";
  }
);

$tftpd->on(
  finish => sub {
    warn "Mojo::TFTPd: finish ./", $_[1]->file, " (", ($_[2] // 'No error'), ")\n";
  }
);

$tftpd->on(
  rrq => sub {
    my ($tftpd, $c) = @_;
    open my $FH, '<', $c->file or return;
    warn "Mojo::TFTPd: rrq ./", $c->file, "\n";
    $c->filehandle($FH);
    $c->filesize(-s $c->file);
  }
);

$tftpd->on(
  wrq => sub {
    my ($tftpd, $c) = @_;
    open my $FH, '>', '/dev/null' or return;
    warn "Mojo::TFTPd: wrq ./", $c->file, "\n";
    $c->filehandle($FH);
  }
);

$tftpd->start;
$tftpd->ioloop->start unless $tftpd->ioloop->is_running;


example/tftp-http-proxy.pl
#!/usr/bin/env perl
use Mojo::Base -strict;
use Mojo::TFTPd;
use Mojo::UserAgent;

package Mojo::TFTPd::Connection::HTTP;
use Mojo::Base 'Mojo::TFTPd::Connection';

sub send_data {
  my $self = shift;
  return 1 if $self->{pause};
  return $self->SUPER::send_data(@_);
}

package main;

my $tftpd = Mojo::TFTPd->new(
  listen           => 'localhost:7000',
  connection_class => 'Mojo::TFTPd::Connection::HTTP',
);

my $ua = Mojo::UserAgent->new;
$ua->proxy->detect;

$tftpd->on(
  rrq => sub {
    my ($tftpd, $c) = @_;
    my $file = $c->file;

    if ($file =~ m!^https?://!) {
      my $tx = $ua->build_tx(GET => $file);
      $c->{pause} = 1;
      $c->filehandle($tx->res->content->asset);
      $tx->res->max_message_size(0);

      Scalar::Util::weaken($c);

      $tx->res->on(
        progress => sub {
          my $msg = shift;

          # do we have enough to send a full chunk?
          my $want = $c->_sequence_number * $c->blocksize;
          my $have = $msg->content->progress;

          if ($want <= $have) {
            delete $c->{pause};
            $c->filehandle($tx->res->content->asset);
            $c->send_data;
            $c->{pause} = 1;
          }
        }
      );

      $ua->start(
        $tx,
        sub {
          my ($ua, $tx) = @_;
          return unless $c;

          # download is complete
          delete $c->{pause};
          $c->filehandle($tx->res->content->asset);
          $c->send_data;
        }
      );
    }
    else {
      ...;
    }
  }
);

$tftpd->start->ioloop->start;


lib/Mojo/TFTPd/Connection.pm
package Mojo::TFTPd::Connection;
use Mojo::Base -base;

use Socket();
use Scalar::Util qw(blessed);

use constant DEBUG        => !!$ENV{MOJO_TFTPD_DEBUG};
use constant OPCODE_DATA  => 3;
use constant OPCODE_ACK   => 4;
use constant OPCODE_ERROR => 5;
use constant OPCODE_OACK  => 6;
use constant ROLLOVER     => 256 * 256;

our %ERROR_CODES = (
  not_defined         => [0, 'Not defined, see error message'],
  unknown_opcode      => [0, 'Unknown opcode: %s'],
  no_connection       => [0, 'No connection'],
  file_not_found      => [1, 'File not found'],
  access_violation    => [2, 'Access violation'],
  disk_full           => [3, 'Disk full or allocation exceeded'],
  illegal_operation   => [4, 'Illegal TFTP operation'],
  unknown_transfer_id => [5, 'Unknown transfer ID'],
  file_exists         => [6, 'File already exists'],
  no_such_user        => [7, 'No such user'],
);

BEGIN {
  # do not use MSG_DONTWAIT on platforms that do not support it (Win32)
  my $msg_dontwait = 0;
  eval { $msg_dontwait = Socket::MSG_DONTWAIT };
  sub MSG_DONTWAIT() {$msg_dontwait}
}

has blocksize        => 512;
has error            => '';
has file             => '/dev/null';
has filehandle       => undef;
has filesize         => undef;
has lastop           => undef;
has mode             => '';
has peerhost         => '';
has peername         => '';
has retransmit       => 0;
has retries          => 2;
has rfc              => sub { +{} };
has socket           => undef;
has timeout          => undef;
has type             => undef;
has _attempt         => 0;
has _sequence_number => 1;

sub receive_ack {
  my $self = shift;
  my ($n)  = unpack 'n', shift;
  my $seq  = $self->_sequence_number % ROLLOVER;

  DEBUG && warn "[Mojo::TFTPd] <<< %s ack %s %s\n", $self->peerhost, $n,
    ($n && $n != $seq ? "expected $seq" : '');

  return $self->send_data if $n == 0 and $self->lastop eq OPCODE_OACK;
  return 0                if $self->lastop eq OPCODE_ERROR;
  return 0 if $self->{last_sequence_number} and $n == $self->{last_sequence_number} % ROLLOVER;

  if ($n == $seq) {
    $self->{_attempt} = 0;
    $self->{_sequence_number}++;
    return $self->send_data;
  }

  return 1                if $self->retransmit and $n < $seq;
  return $self->send_data if $self->{retries}--;
  $self->error('Invalid packet number');
  return 0;
}

sub receive_data {
  my $self = shift;
  my ($n, $data) = unpack 'na*', shift;
  my $seq = $self->_sequence_number % ROLLOVER;

  DEBUG && warn "[Mojo::TFTPd] <<< %s data %s (%s) %s\n", $self->peerhost, $n, length $data,
    ($n != $seq ? " expected $seq" : '');

  unless ($n == $seq) {
    return 1               if $self->retransmit and $n < $seq;
    return $self->send_ack if $self->{retries}--;
    $self->error('Invalid packet number');
    return 0;
  }

  my $handle = $self->filehandle;
  if (blessed $handle and $handle->isa('Mojo::Asset')) {
    local $!;
    eval { $handle->add_chunk($data) };
    return $self->send_error(illegal_operation => "Unable to add chunk $!") if $!;
  }
  elsif (!$handle->syswrite($data)) {
    return $self->send_error(illegal_operation => "Write: $!");
  }

  unless (length $data == $self->blocksize) {
    $self->{last_sequence_number} = $n;
  }

  return $self->send_error(disk_full => 'tsize exceeded')
    if $self->filesize and $self->filesize < $self->blocksize * ($n - 1) + length $data;

  $self->{_sequence_number}++;
  return $self->send_ack;
}

sub receive_error {
  my $self = shift;
  my ($code, $msg) = unpack 'nZ*', shift;

  warn "[Mojo::TFTPd] <<< $self->{peerhost} error $code $msg\n" if DEBUG;
  $self->error("($code) $msg");
  return 0;
}

sub send_ack {
  my $self = shift;
  $self->{lastop} = OPCODE_ACK;

  my $seq = ($self->_sequence_number - 1) % ROLLOVER;
  DEBUG && warn "[Mojo::TFTPd] <<< %s ack %s %s\n", $self->peerhost, $seq,
    ($self->_attempt ? " retransmit $self->{_attempt}" : '');

  my $sent = $self->socket->send(pack('nn', OPCODE_ACK, $seq), MSG_DONTWAIT, $self->peername);
  return 0 if defined $self->{last_sequence_number};
  return 1 if $sent or $self->{retries}--;
  $self->error("Send: $!");
  return 0;
}

sub send_data {
  my $self = shift;
  $self->{lastop} = OPCODE_DATA;

  my ($handle, $n, $data) = ($self->filehandle, $self->_sequence_number);
  if (blessed $handle and $handle->isa('Mojo::Asset')) {
    $data = $handle->get_chunk(($n - 1) * $self->blocksize, $self->blocksize);
    return $self->send_error(file_not_found => 'Unable to read chunk') unless defined $data;
  }
  elsif (not seek $handle, ($n - 1) * $self->blocksize, 0) {
    return $self->send_error(file_not_found => "Seek: $!");
  }
  elsif (not defined $handle->sysread($data, $self->blocksize)) {
    return $self->send_error(file_not_found => "Read: $!");
  }

  if (length $data < $self->blocksize) {
    $self->{last_sequence_number} = $n;
  }

  my $seq = $n % ROLLOVER;
  DEBUG && warn sprintf "[Mojo::TFTPd] >>> %s data %s (%s) %s\n", $self->{peerhost}, $seq,
    length $data, $self->_attempt ? "retransmit $self->{_attempt}" : '';

  my $sent
    = $self->socket->send(pack('nna*', OPCODE_DATA, $seq, $data), MSG_DONTWAIT, $self->peername);

  return 0 unless length $data;
  return 1 if $sent or $self->{retries}--;
  $self->error("Send: $!");
  return 0;
}

sub send_error {
  my ($self, $name) = @_;
  my $err = $ERROR_CODES{$name} || $ERROR_CODES{not_defined};

  $self->{lastop} = OPCODE_ERROR;
  warn "[Mojo::TFTPd] >>> $self->{peerhost} error @$err\n" if DEBUG;

  $self->error($_[2]);
  $self->socket->send(pack('nnZ*', OPCODE_ERROR, @$err), MSG_DONTWAIT, $self->peername);

  return 0;
}

sub send_oack {
  my $self = shift;
  $self->{lastop} = OPCODE_OACK;

  my @options;
  push @options, 'blksize', $self->blocksize if $self->rfc->{blksize};
  push @options, 'timeout', $self->timeout   if $self->rfc->{timeout};
  push @options, 'tsize',   $self->filesize  if exists $self->rfc->{tsize} and $self->filesize;

  warn "[Mojo::TFTPd] >>> $self->{peerhost} oack @options"
    . ($self->_attempt ? " retransmit $self->{_attempt}" : '') . "\n"
    if DEBUG;

  my $sent = $self->socket->send(pack('na*', OPCODE_OACK, join "\0", @options),
    MSG_DONTWAIT, $self->peername);
  return 1 if $sent or $self->{retries}--;
  $self->error("Send: $!");
  return 0;
}

sub send_retransmit {
  my $self = shift;
  return 0 unless $self->lastop;

  unless ($self->retransmit) {
    $self->error('Inactive timeout');
    return 0;
  }

  # Errors are not retransmitted
  return 0 if $self->lastop == OPCODE_ERROR;

  if ($self->_attempt >= $self->retransmit) {
    $self->error('Retransmit timeout');
    return 0;
  }

  $self->{_attempt}++;

  return $self->send_oack if $self->lastop eq OPCODE_OACK;
  return $self->send_ack  if $self->lastop eq OPCODE_ACK;
  return $self->send_data if $self->lastop eq OPCODE_DATA;
  return 0;
}

1;

=encoding utf8

=head1 NAME

Mojo::TFTPd::Connection - A connection class for Mojo::TFTPd

=head1 SYNOPSIS

See L<Mojo::TFTPd>

=head1 ATTRIBUTES

=head2 type

  $str = $connection->type;

Type of connection rrq or wrq

=head2 blocksize

  $int = $connection->blocksize;

The negotiated blocksize. Default is 512 Byte.

=head2 error

  $str = $connection->error;

Useful to check inside L<Mojo::TFTPd/finish> events to see if anything has
gone wrong. Holds a string describing the error.

=head2 file

  $str = $connection->file;

The filename the client requested to read or write.

=head2 filehandle

  $fh = $connection->filehandle;

This must be set inside the L<rrq|Mojo::TFTPd/rrq> or L<wrq|Mojo::TFTPd/wrq>
event or the connection will be dropped.
Can be either L<Mojo::Asset> or filehandle.

=head2 filesize

  $int = $connection->filesize;

This must be set inside the L<rrq|Mojo::TFTPd/rrq>
to report "tsize" option if client requested.

If set inside L<wrq|Mojo::TFTPd/wrq> limits maximum upload size.
Set automatically on WRQ with "tsize" option.

Can be used inside L<finish|Mojo::TFTPd/finish> for uploads
to check if reported "tsize" and received data length match.

=head2 timeout

  $num = $connection->timeout;

Retransmit/Inactive timeout.

=head2 lastop

  $str = $connection->lastop;

Last operation.

=head2 mode

  $str = $connection->mode;

Either "netascii", "octet" or empty string if unknown.

=head2 peerhost

  $str = $connection->peerhost;

The IP address of the remote client.

=head2 peername

  $bin = $connection->peername;

Packet address of the remote client.

=head2 retries

  $int = $connection->retries;

Number of times L</send_data>, L</send_ack> or L</send_oack> can be retried before the
connection is dropped.
This value comes from L<Mojo::TFTPd/retries> or set inside L<rrq|Mojo::TFTPd/rrq> or L<wrq|Mojo::TFTPd/wrq>
events.

=head2 retransmit

  $int = $connection->retransmit;

Number of times last operation (L</send_data>, L</send_ack> or L</send_oack>)
to be retransmitted on timeout before the connection is dropped.
This value comes from L<Mojo::TFTPd/retransmit> or set inside L<rrq|Mojo::TFTPd/rrq> or L<wrq|Mojo::TFTPd/wrq>
events.

Retransmits are disabled if set to 0.

=head2 socket

  $fh = $connection->socket;

The UDP handle to send data to.

=head2 rfc

  $hash_ref = $connection->rfc;

Contains RFC 2347 options the client has provided.

=head1 METHODS

=head2 receive_ack

  $bool = $connection->receive_ack($bytes);

This method is called when the client sends ACK to the server.

=head2 receive_data

  $bool = $connection->receive_data($bytes);

This method is called when the client sends DATA to the server.

=head2 receive_error

  $bool = $connection->receive_error($bytes);

This method is called when the client sends ERROR to the server.

=head2 send_ack

  $bool = $connection->send_ack;

This method is called when the server sends ACK to the client.

=head2 send_data

  $bool = $connection->send_data;

This method is called when the server sends DATA to the client.

=head2 send_error

  $bool = $connection->send_error($key => $descr);

Used to report error to the client.

=head2 send_oack

  $bool = $connection->send_oack;

Used to send RFC 2347 OACK to client

Supported options are

=over

=item RFC 2348 blksize

Report L</blocksize>.

=item RFC 2349 timeout

Report L</timeout>.

=item RFC 2349 tsize

Report L</filesize> if set inside the L<rrq|Mojo::TFTPd/rrq>.

=back

=head2 send_retransmit

  $bool = $connection->send_retransmit;

Used to retransmit last packet to the client.

=head1 SEE ALSO

L<Mojo::TFTPd>

=cut


example/custom-connection-class.pl
#!/usr/bin/env perl
use Mojo::Base -strict;
use Mojo::TFTPd;
use Mojo::UserAgent;

{

  package Mojo::TFTPd::Connection::HTTP;
  use Mojo::Base 'Mojo::TFTPd::Connection';

  sub send_data {
    my $self = shift;
    return 1 if $self->{pause};
    return $self->SUPER::send_data(@_);
  }
}

package main;
my $tftpd = Mojo::TFTPd->new(listen => 'localhost:7000',
  connection_class => 'Mojo::TFTPd::Connection::HTTP');
my $ua = Mojo::UserAgent->new;

$tftpd->on(
  rrq => sub {
    my ($tftpd, $c) = @_;
    my $file = $c->file;

    if ($file =~ m!^https?://!) {
      my $tx = $ua->build_tx(GET => $file);

      $c->{pause} = 1;
      $c->filehandle($tx->res->content->asset);
      $tx->res->max_message_size(0);

      Scalar::Util::weaken($c);
      $ua->start(
        $tx,
        sub {
          my ($ua, $tx) = @_;
          return unless $c;
          delete $c->{pause};
          $c->filehandle($tx->res->content->asset);
          $c->send_data;
        }
      );
    }
    else {
      # ...
    }
  }
);

$tftpd->start->ioloop->start;


lib/Mojo/TFTPd.pm
package Mojo::TFTPd;
use Mojo::Base 'Mojo::EventEmitter';

use Mojo::IOLoop;
use Mojo::TFTPd::Connection;
use Scalar::Util qw(weaken);

use constant DEBUG          => !!$ENV{MOJO_TFTPD_DEBUG};
use constant OPCODE_RRQ     => 1;
use constant OPCODE_WRQ     => 2;
use constant OPCODE_DATA    => 3;
use constant OPCODE_ACK     => 4;
use constant OPCODE_ERROR   => 5;
use constant OPCODE_OACK    => 6;
use constant MIN_BLOCK_SIZE => 8;
use constant MAX_BLOCK_SIZE => 65464;                      # From RFC 2348

our $VERSION = '0.05';

has connection_class   => 'Mojo::TFTPd::Connection';
has inactive_timeout   => 15;
has ioloop             => sub { Mojo::IOLoop->singleton };
has listen             => 'tftp://*:69';
has max_connections    => 1000;
has retransmit         => 0;
has retransmit_timeout => 2;
has retries            => 1;

sub start {
  my $self = shift;
  return $self if $self->{connections};

  # split $self->listen into host and port
  my ($host, $port) = $self->_parse_listen;
  warn "[Mojo::TFTPd] Listen to $host:$port\n" if DEBUG;

  my $socket = IO::Socket::INET->new(LocalAddr => $host, LocalPort => $port, Proto => 'udp');
  return $self->emit(error => "Can't create listen socket: $!") unless $socket;

  my $reactor = $self->ioloop->reactor;
  weaken $self;
  $socket->blocking(0);
  $reactor->io($socket, sub { $self->_incoming });
  $reactor->watch($socket, 1, 0);    # watch read events
  $self->{connections} = {};
  $self->{socket}      = $socket;

  return $self;
}

sub _delete_connection {
  my ($self, $connection) = @_;
  delete $self->{connections}{$connection->peername};
  $self->ioloop->remove($connection->{timer}) if $connection->{timer};
  $self->emit(finish => $connection, $connection->error);
}

sub _incoming {
  my $self   = shift;
  my $socket = $self->{socket};

  # Add 4 Bytes of Opcode + Block#
  return $self->emit(error => "Read: $!")
    unless defined(my $read = $socket->recv(my $datagram, MAX_BLOCK_SIZE + 4));

  my $opcode = unpack 'n', substr $datagram, 0, 2, '';
  if ($opcode eq OPCODE_RRQ) {
    return $self->_new_request(rrq => $datagram);
  }
  elsif ($opcode eq OPCODE_WRQ) {
    return $self->_new_request(wrq => $datagram);
  }

  my $connection = $self->{connections}{$socket->peername};
  return $self->emit(error => "@{[$socket->peerhost]} has no connection") unless $connection;

  if ($opcode == OPCODE_ACK) {
    $connection->receive_ack($datagram)
      ? $self->_reset_timer($connection)
      : $self->_delete_connection($connection);
  }
  elsif ($opcode == OPCODE_DATA) {
    $connection->receive_data($datagram)
      ? $self->_reset_timer($connection)
      : $self->_delete_connection($connection);
  }
  elsif ($opcode == OPCODE_ERROR) {
    $connection->receive_error($datagram);
    $self->_delete_connection($connection);
  }
  else {
    $connection->error('Unknown opcode');
    $self->_delete_connection($connection);
  }
}

sub _new_request {
  my ($self, $type, $datagram) = @_;
  my ($file, $mode, @rfc) = split "\0", $datagram;
  my $socket = $self->{socket};
  warn "[Mojo::TFTPd] <<< @{[$socket->peerhost]} new request $type $file $mode @rfc\n" if DEBUG;

  return $self->emit(error => "Cannot handle $type requests") unless $self->has_subscribers($type);
  return $self->emit(error => "Max connections ($self->{max_connections}) reached")
    if $self->max_connections <= keys %{$self->{connections}};

  my %rfc        = @rfc;
  my $connection = $self->connection_class->new(
    type       => $type,
    file       => $file,
    mode       => $mode,
    peerhost   => $socket->peerhost,
    peername   => $socket->peername,
    retries    => $self->retries,
    timeout    => $self->retransmit ? $self->retransmit_timeout : $self->inactive_timeout,
    retransmit => $self->retransmit,
    rfc        => \%rfc,
    socket     => $socket,
  );

  if ($rfc{blksize}) {
    $rfc{blksize} = MIN_BLOCK_SIZE if $rfc{blksize} < MIN_BLOCK_SIZE;
    $rfc{blksize} = MAX_BLOCK_SIZE if $rfc{blksize} > MAX_BLOCK_SIZE;
    $connection->blocksize($rfc{blksize});
  }
  if ($rfc{timeout} and $rfc{timeout} >= 0 and $rfc{timeout} <= 255) {
    $connection->timeout($rfc{timeout});
  }
  if ($type eq 'wrq' and $rfc{tsize}) {
    $connection->filesize($rfc{tsize});
  }

  $self->emit($type => $connection);

  if (!$connection->filehandle) {
    $connection->send_error(file_not_found => $connection->error || 'No filehandle');
    $self->_reset_timer($connection);
  }
  elsif (%rfc and $connection->send_oack) {
    $self->{connections}{$connection->peername} = $connection;
    $self->_reset_timer($connection);
  }
  elsif ($type eq 'rrq' ? $connection->send_data : $connection->send_ack) {
    $self->{connections}{$connection->peername} = $connection;
    $self->_reset_timer($connection);
  }
  else {
    $self->emit(finish => $connection, $connection->error);
  }
}

sub _parse_listen {
  my $self = shift;

  my ($scheme, $host, $port) = $self->listen =~ m!
      (?: ([^:/]+) :// )?   # part before ://
      ([^:]*)               # everyting until a :
      (?: : (\d+) )?        # any digits after the :
    !xms;

  $port = getservbyname($scheme, '') if $scheme && !defined $port;
  $port //= 69;
  $host = '0.0.0.0' if $host eq '*';

  return ($host, $port);
}

sub _reset_timer {
  my ($self, $connection) = @_;

  $self->ioloop->remove($connection->{timer}) if $connection->{timer};
  $connection->{timer} = $self->ioloop->recurring(
    $connection->timeout,
    sub {
      $connection->send_retransmit or $self->_delete_connection($connection);
    }
  );
}

sub DESTROY {
  return if ${^GLOBAL_PHASE} eq 'DESTRUCT';
  my $self = shift;
  $self->ioloop->reactor->remove($self->{socket}) if $self->{socket};
}

1;

=encoding utf8

=head1 NAME

Mojo::TFTPd - Trivial File Transfer Protocol daemon

=head1 VERSION

0.04

=head1 SYNOPSIS

  use Mojo::TFTPd;
  my $tftpd = Mojo::TFTPd->new;

  $tftpd->on(error => sub ($tftpd, $error) { warn "TFTPd: $error\n" });

  $tftpd->on(rrq => sub ($tftpd, $connection) {
    open my $FH, '<', $connection->file;
    $connection->filehandle($FH);
    $connection->filesize(-s $connection->file);
  });

  $tftpd->on(wrq => sub ($tftpd, $connection) {
    open my $FH, '>', '/dev/null';
    $connection->filehandle($FH);
  });

  $tftpd->on(finish => sub ($tftpd, $connection, $error) {
    warn "Connection: $error\n" if $error;
  });

  $tftpd->start;
  $tftpd->ioloop->start unless $tftpd->ioloop->is_running;

=head1 DESCRIPTION

This module implements a server for the
L<Trivial File Transfer Protocol|http://en.wikipedia.org/wiki/Trivial_File_Transfer_Protocol>.

From Wikipedia:

  Trivial File Transfer Protocol (TFTP) is a file transfer protocol notable
  for its simplicity. It is generally used for automated transfer of
  configuration or boot files between machines in a local environment.

The connection which is referred to in this document is an instance of
L<Mojo::TFTPd::Connection>.

=head1 EVENTS

=head2 error

  $tftpd->on(error => sub ($tftpd, $str) { ... });

This event is emitted when something goes wrong: Fail to L</listen> to socket,
read from socket or other internal errors.

=head2 finish

  $tftpd->on(finish => sub ($tftpd, $connection, $error) { ... });

This event is emitted when the L<Mojo::TFTPd::Connection> finish, either
successfully or due to an error. C<$error> will be an empty string on success.

=head2 rrq

  $tftpd->on(rrq => sub ($tftpd, $connection) { ... });

This event is emitted when a new read request arrives from a client. The
callback should set L<Mojo::TFTPd::Connection/filehandle> or the connection
will be dropped.
L<Mojo::TFTPd::Connection/filehandle> can also be a L<Mojo::Asset> reference.

=head2 wrq

  $tftpd->on(wrq => sub ($tftpd, $connection) { ... });

This event is emitted when a new write request arrives from a client. The
callback should set L<Mojo::TFTPd::Connection/filehandle> or the connection
will be dropped.
L<Mojo::TFTPd::Connection/filehandle> can also be a L<Mojo::Asset> reference.

=head1 ATTRIBUTES

=head2 connection_class

  $str = $tftpd->connection_class;
  $tftpd = $tftpd->connection_class($str);

Used to set a custom connection class. Defaults to L<Mojo::TFTPd::Connection>.

=head2 inactive_timeout

  $num = $tftpd->inactive_timeout;
  $tftpd = $tftpd->inactive_timeout(15);

How long a L<connection|Mojo::TFTPd::Connection> can stay idle before
being dropped. Default is 15 seconds.

=head2 ioloop

  $loop = $tftpd->ioloop;
  $tftpd = $tftpd->ioloop(Mojo::IOLoop->new);

Holds an instance of L<Mojo::IOLoop>.

=head2 listen

  $str = $tftpd->listen;
  $tftpd = $tftpd->listen('127.0.0.1:69');
  $tftpd = $tftpd->listen('tftp://*:69');

The bind address for this server.

=head2 max_connections

  $int = $tftpd->max_connections;
  $tftpd = $tftpd->max_connections(1000);

How many concurrent connections this server can handle. Default to 1000.

=head2 retransmit

  $int = $tftpd->retransmit;
  $tftpd = $tftpd->retransmit(1);

How many times the server should try to retransmit the last packet on timeout before
dropping the L<connection|Mojo::TFTPd::Connection>. Default is 0 (disable retransmits)

=head2 retransmit_timeout

  $num = $tftpd->retransmit_timeout;
  $tftpd = $tftpd->retransmit_timeout(2);

How long a L<connection|Mojo::TFTPd::Connection> can stay idle before last packet
being retransmitted. Default is 2 seconds.

=head2 retries

  $int = $tftpd->retries;
  $tftpd = $tftpd->retries(1);

How many times the server should try to send ACK or DATA to the client before
dropping the L<connection|Mojo::TFTPd::Connection>.

=head1 METHODS

=head2 start

  $tftpd = $tftpd->start;

Starts listening to the address and port set in L</Listen>. The L</error>
event will be emitted if the server fail to start.

=head1 AUTHORS

Jan Henning Thorsen - C<jhthorsen@cpan.org>

Per Carlson - C<pelle@hemmop.com>

Philippe Bruhat (BooK) - C<book@cpan.org>

Svetoslav Naydenov - C<harryl@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=cut
